grammar/abnf-grammar.txt

```diff
@@ -542,7 +542,7 @@ symbol = "!" / "&&" / "||"
        / "[" / "]"
        / "{" / "}"
        / "," / "." / ".." / "..." / ";" / ":" / "::" / "?"
-       / "->" / "_"
+       / "->" / "_" / "$"
```

Comment:
In the ABNF grammar, I think that there is currently no need need to have `$` as a separate token, because `$` can only appear at the start of an `address-literal` token, and nowhere else in Leo.

Note that the rule for address literals could be equivalently stated as
```
address-literal = %s"$aleo1" 58( lowercase-letter / digit )
```
(nothing wrong with your formulation; just equivalent).

Since these rules are part of the lexical subgrammar, no whitespace or comments are allowed between `$` and `aleo1...`, i.e. `$aleo`...` has to be a single token -- at least in the grammar.

Reply:
@acoglio The reason I put it as a separate token was to align it more with the parser. Otherwise, we would be parsing by $aleo1 at the start of identifiers, which is no different from how we originally parsed them. My understanding in part of this change was to identify $ and then parse an address-literal directly. As for spacing, I will double-check no spaces between $ and aleo1 and I should add a test for that case.

Reply:
I guess I was thinking of a slightly different tokenization approach; let me explain.

If I understand the latest code (i.e. commit 82981ba) correctly (please correct me if I'm wrong), it does the following:
- It tokenizes `$aleo1...` as two tokens, namely `$` and `aleo1...`. The latter is an `AddressLit` token due to the initial test in https://github.com/AleoHQ/leo/blob/82981ba2166133b13d3eaf47e0a8a16b5d87a013/parser/src/tokenizer/lexer.rs#L310
- The post-tokenization parsing code then ensures that there is no whitespace between the two tokens (which you added in that commit).

Here's what I was thinking instead:
- Adding an `eat_address_literal` method that attempts to read a contiguous `$aleo1<58moreLowercaseLettersOrDigits>` as one token, accordingly with the ABNF grammar definition of a token as, among other things, an address literal.
- Modifying https://github.com/AleoHQ/leo/blob/82981ba2166133b13d3eaf47e0a8a16b5d87a013/parser/src/tokenizer/lexer.rs#L179 to call `eat_address_literal`, i.e. attempt to read an address when encountering `$`.
- Avoiding the test in https://github.com/AleoHQ/leo/blob/82981ba2166133b13d3eaf47e0a8a16b5d87a013/parser/src/tokenizer/lexer.rs#L310
- Avoiding the post-tokenization parsing test that there is no whitespace between the two tokens.
- Avoiding `"$"` as a symbol token in the ABNF grammar.

Reply:
This is just an idea. There is of course nothing wrong with the current implementation. But if a goal is to be compatible with the ABNF grammar, then the fact that `$aleo1...` is parsed as two tokens `$` and `aleo1...` is already incompatible with the grammar definition of `token --> atomic-literal --> address-literal --> $aleo1...`, where `$aleo1...` is thus a single token.

Reply:
Saw the latest commit -- looks good, thanks.

Reply:
@acoglio, I went ahead and made the suggested changes. $ is no longer a token. Though I didn't make a separate function for eating address literals, the logic would be the same as eating an identifier.